lib/azure/armrest/model/hash_model.rb

```diff
@@ -0,0 +1,138 @@
+require 'active_support/core_ext/string/inflections'
+
+module Azure
+  module Armrest
+    class HashModel
+      class << self
+        private
+
+        # Convert a key to a model class attribute which must be a qualified ruby method name
+        # key is from the original JSON or Hash object
+        #
+        # This is the default implementation, to be overridden by subclasses.
+        #
+        def key_to_attribute(key)
+          key.to_s.underscore
+        end
+
+        # Should setter methods be automatically created?
+        #
+        # Default to true, to be overridden by subclasses.
+        #
+        def generate_writer?
+          true
+        end
+
+        # Should values of this attribute be excluded from converting to model objects
+        #
+        # Default to false, to be overridden by subclasses.
+        #
+        def attr_excluded?(_attr_name)
+          false
+        end
+
+        # Declare attributes that need to be converted to Model
+        def attr_model(*attrs)
+          # Merge the declared attributes to the modelized list
+          @modelized_list = modelized_list | Set.new(attrs.map(&:to_s))
+        end
+
+        # Return the list of modelized attributes
+        def modelized_list
+          @modelized_list ||= Set.new
+        end
+
+        def convert_value(key, value, parent)
+          method = key_to_attribute(key)
+          new_val =
+            if attr_excluded?(method)
+              value
+            else
+              # Must deal with nested models
+              case value
+              when Array
+                value.collect do |elem|
+                  elem.kind_of?(Hash) ? parent.send(:hash_to_model, method.camelize.singularize, elem) : elem
+                end
+              when Hash
+                parent.send(:hash_to_model, method.camelize, value)
+              else
+                value
+              end
+            end
+
+          add_accessor_methods(method, key)
+          new_val
+        end
+
+        def add_accessor_methods(method, key)
+          method = "_#{method}" if instance_methods.include?(method.to_sym)
+          class_eval { define_method(method) { @data[key] } }
```

Comment:
So, if I understand this correctly, won't every model end up getting the `_method` plus the original eventually, given enough objects have been instantiated?

The first time we go through this method, it would work as expected (assuming something like `#tags` or the like isn't already defined), and define the `"#{method}"`.  But the second time through, we would then end up defining the `"_#{method}"` because there is nothing stopping this from being redefined otherwise.

Seems like we need some kind of cache in place to check to see if we have defined methods using `add_accessor_methods` prior to doing it again because there is really nothing stoping use from entering this method again after we have run through it the first time with the same args.

Reply:
Oops, my bad. I was trying to solve the case issue.